go/vt/vtgate/engine/route.go

```diff
@@ -425,101 +425,178 @@ func (route *Route) paramsAllShards(vcursor VCursor, bindVars map[string]*queryp
+		if err != nil {
+			log.Errorf("failed to route information_schema query to keyspace [%s]", ks)
+			bindVars[sqltypes.BvSchemaName] = sqltypes.StringBindVariable(ks)
+			return defaultRoute(bindVars)
+		}
+		setReplaceSchemaName(bindVars)
+		return destinations, []map[string]*querypb.BindVariable{bindVars}, nil
 	}
 
 	if len(route.SysTableTableName) == 0 && len(route.SysTableTableSchema) == 0 {
-		return defaultRoute()
+		return defaultRoute(bindVars)
 	}
 
 	env := evalengine.ExpressionEnv{
 		BindVars: bindVars,
 		Row:      []sqltypes.Value{},
 	}
 
-	var specifiedKS string
+	specifiedKSs := make([]string, 0)
 	for _, tableSchema := range route.SysTableTableSchema {
 		result, err := tableSchema.Evaluate(env)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		ks := result.Value().ToString()
-		if specifiedKS == "" {
-			specifiedKS = ks
+		containsDup := false
+		for _, existingKs := range specifiedKSs {
+			if existingKs == ks {
+				containsDup = true
+				break
+			}
 		}
-		if specifiedKS != ks {
-			return nil, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "specifying two different database in the query is not supported")
+		if !containsDup {
+			specifiedKSs = append(specifiedKSs, ks)
 		}
 	}
-	if specifiedKS != "" {
-		bindVars[sqltypes.BvSchemaName] = sqltypes.StringBindVariable(specifiedKS)
-	}
 
-	var tableName string
+	tableNames := make([]string, 0)
 	for _, sysTableName := range route.SysTableTableName {
 		val, err := sysTableName.Evaluate(env)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		tabName := val.Value().ToString()
-		if tableName == "" {
-			tableName = tabName
+		containsDup := false
+		for _, existingTabName := range tableNames {
+			if existingTabName == tabName {
+				containsDup = true
+				break
+			}
 		}
-		if tableName != tabName {
-			return nil, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "two predicates for table_name not supported")
+		if !containsDup {
+			tableNames = append(tableNames, tabName)
 		}
 	}
-	if tableName != "" {
-		bindVars[BvTableName] = sqltypes.StringBindVariable(tableName)
-	}
-
-	// if the table_schema is system system, route to default keyspace.
-	if sqlparser.SystemSchema(specifiedKS) {
-		return defaultRoute()
-	}
-
-	// the use has specified a table_name - let's check if it's a routed table
-	if tableName != "" {
-		rss, err := route.paramsRoutedTable(vcursor, bindVars, specifiedKS, tableName)
-		if err != nil {
-			// Only if keyspace is not found in vschema, we try with default keyspace.
-			// As the in the table_schema predicates for a keyspace 'ks' it can contain 'vt_ks'.
-			if vterrors.ErrState(err) == vterrors.BadDb {
-				return defaultRoute()
+	if len(tableNames) == 0 {
+		tableNames = append(tableNames, "")
+	}
+
+	bvs := make([]map[string]*querypb.BindVariable, 0)
+	shards := make([]*srvtopo.ResolvedShard, 0)
+
+	for _, specifiedKS := range specifiedKSs {
+		for _, tableName := range tableNames {
```

Comment:
There are at least two issues with your solution.
1.  The logic is correct only for in clause, but for union clause and join clause. this may have issue, For example, we may have a schema A table C, D and schema B, table C, D. When user input a join of A.C join B.D, your code will return all A.c, A.D, B.C, B.D
2. This part of logic partition each schema and table into different shards. But most likely they will belong to the same shard, so assume you have a M schema and N tables. you will executed M*N queries. Although they may all belong to the same shard.